snappy/snap_local.go

```diff
@@ -57,87 +59,115 @@ func NewInstalledSnap(yamlPath string) (*Snap, error) {
 	return snap, nil
 }
 
+func completeInfo(info *snap.Info, manifest *diskManifest) {
+	// XXX: we actually should not trust snap.yaml name!
+
+	info.Developer = SideloadedDeveloper
+	// default for compat with older installs
+	info.Channel = "stable"
+	info.Size = -1
+
+	if manifest != nil {
+		info.Revision = manifest.Revision
+		info.Developer = manifest.Developer
+		info.Channel = manifest.Channel
+		// store edits win!
+		info.Description = manifest.Description
+		// XXX: Summary doesn't exist in the store yet anyway
+		info.Size = manifest.Size
+	}
+}
+
 // newSnapFromYaml returns a new Snap from the given *snapYaml at yamlPath
 func newSnapFromYaml(yamlPath string, m *snapYaml) (*Snap, error) {
-	snap := &Snap{
+	s := &Snap{
 		basedir: filepath.Dir(filepath.Dir(yamlPath)),
 		m:       m,
 	}
 
-	// override the package's idea of its version
-	// because that could have been rewritten on sideload
-	// and developer is empty sideloaded ones.
-	m.Version = filepath.Base(snap.basedir)
-
 	// check if the snap is active
-	allVersionsDir := filepath.Dir(snap.basedir)
+	allVersionsDir := filepath.Dir(s.basedir)
 	p, err := filepath.EvalSymlinks(filepath.Join(allVersionsDir, "current"))
 	if err != nil && !os.IsNotExist(err) {
 		return nil, err
 	}
 
-	if p == snap.basedir {
-		snap.isActive = true
+	if p == s.basedir {
+		s.isActive = true
+	}
+
+	// XXX: temp ugly hack for now
+	var yamlBits []byte
+	if osutil.FileExists(yamlPath) {
+		yamlBits, err = ioutil.ReadFile(yamlPath)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// XXX: ugly serialize and reparse
+		yamlBits, err = yaml.Marshal(m)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	manifestPath := ManifestPath(snap.Info())
+	info, err := snap.InfoFromSnapYaml(yamlBits)
+	if err != nil {
+		return nil, err
+	}
+	s.info = info
+
+	var manifest *diskManifest
+	manifestPath := ManifestPath(info)
 	if osutil.FileExists(manifestPath) {
 		content, err := ioutil.ReadFile(manifestPath)
 		if err != nil {
 			return nil, err
 		}
 
-		var manifest diskManifest
-		if err := yaml.Unmarshal(content, &manifest); err != nil {
+		manifest = new(diskManifest)
+		if err := yaml.Unmarshal(content, manifest); err != nil {
 			return nil, &ErrInvalidYaml{File: manifestPath, Err: err, Yaml: content}
 		}
-		snap.manifest = &manifest
 	}
 
-	return snap, nil
+	completeInfo(info, manifest)
+
+	// XXX: FIXME: just some tests need this atm
+	// override the package's idea of its version
+	// because that could have been rewritten on sideload
+	// and developer is empty sideloaded ones.
```

Comment:
I don't understand what this was supposed to do, but it feels like it doesn't make much sense in the modern world.